File: 卡码/代码随想录/1/09-动态规划/C++/dp/02-背包问题/src/02-完全背包/Kama_0052.cpp
//52. 携带研究材料
#include <iostream>
#include <vector>
using namespace std;

int main(){
	int n,v;
	scanf("%d %d",&n,&v);
	int weight[n];
	int value[n];
	for(int i=0;i<n;i++){
		scanf("%d %d",&weight[i],&value[i]);
	}
	vector<int> dp(v+1,0);
	for(int i=0;i<n;i++){
		for(int j=weight[i];j<=v;j++){
			dp[j]=max(dp[j],dp[j-weight[i]]+value[i]);
		}
	}
	printf("%d\n",dp[v]);
  return 0;
}

File: 卡码/代码随想录/Java/codeThoughts/src/main/java/com/kurumi/leetCode/codeThoughts/动态规划/_02_背包问题/_02_完全背包/Kama_0052.java
package com.kurumi.leetCode.codeThoughts.动态规划._02_背包问题._02_完全背包;

import java.util.Scanner;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 52. 携带研究材料
 */
public class Kama_0052 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int v = scanner.nextInt();
        int[] weight = new int[n];
        int[] value = new int[n];
        for (int i = 0; i < n; i++) {
            weight[i] = scanner.nextInt();
            value[i] = scanner.nextInt();
        }
        int[] dp = new int[v + 1];
        for (int i = 0; i < n; i++) {
            for (int j = weight[i]; j <= v; j++) {
                dp[j] = Math.max(dp[j], dp[j - weight[i]] + value[i]);
            }
        }
        scanner.close();
        System.out.println(dp[v]);
    }
}
